core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -135,20 +144,33 @@ public void onNext(HttpObject o) {
 
                 final ResponseHeaders headers = (ResponseHeaders) o;
                 final HttpStatus status = headers.status();
+                final ResponseHeaders composedHeaders;
                 if (status.isInformational()) {
-                    // Needs non-informational headers.
-                    break;
-                }
-
-                if (req.method() == HttpMethod.HEAD) {
-                    endOfStream = true;
-                } else if (status.isContentAlwaysEmpty()) {
-                    state = State.NEEDS_TRAILERS;
+                    if (endOfStream) {
+                        failAndRespond(new IllegalStateException(
+                                "published an informational headers whose endOfStream is true: " + o +
+                                " (service: " + service() + ')'));
+                        return;
+                    }
+                    composedHeaders = headers;
                 } else {
-                    state = State.NEEDS_DATA_OR_TRAILERS;
+                    if (req.method() == HttpMethod.HEAD) {
+                        endOfStream = true;
+                    } else if (status.isContentAlwaysEmpty()) {
+                        state = State.NEEDS_TRAILERS;
+                    } else {
+                        state = State.NEEDS_DATA_OR_TRAILERS;
+                    }
+                    if (endOfStream) {
+                        setDone();
+                    }
+                    composedHeaders = composeResponseHeaders(headers, reqCtx.additionalResponseHeaders());
+                    logBuilder().responseHeaders(composedHeaders);
                 }
 
-                logBuilder().responseHeaders(headers);
+                responseEncoder.writeHeaders(req.id(), req.streamId(), composedHeaders, endOfStream,
+                                             reqCtx.additionalResponseTrailers().isEmpty())
+                               .addListener(new WriteHeadersFutureListener(endOfStream));
```

Comment:
Probably better reusing a single instance for the case where `endOfStream` is `false`?

Reply:
(.. because a user can send many informational headers)

Reply:
Yeah, I think it's better too. Also, added for streaming data.